text/0000-repr-align.md

```diff
@@ -0,0 +1,104 @@
+- Feature Name: `repr_align`
+- Start Date: 2015-11-09
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Extend the existing `#[repr]` attribute on structs with an `align = "N"` option
+to specify a custom alignment for `struct` types.
+
+# Motivation
+[motivation]: #motivation
+
+The alignment of a type is normally not worried about as the compiler will "do
+the right thing" of picking an appropriate alignment for general use cases.
+There are situations, however, where a nonstandard alignment may be desired when
+operating with foreign systems. For example these sorts of situations tend to
+necessitate or be much easier with a custom alignment:
+
+* Hardware can often have obscure requirements such as "this structure is
+  aligned to 32 bytes" when it in fact is only composed of 4-byte values. While
+  this can typically be manually calculated and managed, it's often also useful
+  to express this as a property of a type to get the compiler to do a little
+  extra work instead.
+* C compilers like gcc and clang offer the ability to specify a custom alignment
+  for structures, and Rust can much more easily interoperate with these types if
+  Rust can also mirror the request for a custom alignment (e.g. passing a
+  structure to C correctly is much easier).
+* Custom alignment can often be used for various tricks here and there and is
+  often convenient as "let's play around with an implementation" tool. For
+  example this can be used to statically allocate page tables in a kernel
+  or create an at-least cache-line-sized structure easily for concurrent
+  programming.
+
+Currently these sort of situations are possible in Rust but aren't necessarily
+the most ergonomic as programmers must manually manage alignment. The purpose of
+this RFC is to provide a lightweight annotation to alter the compiler-inferred
+alignment of a structure to enable these situations much more easily.
+
+# Detailed design
+[design]: #detailed-design
+
+The `#[repr]` attribute on `struct`s will be extended to include a form such as:
+
+```rust
+#[repr(align = "16")]
+struct MoreAligned(i32);
+```
+
+This structure will still have an alignment of 16 (as returned by
+`mem::align_of`), and in this case the size will also be 16.
+
+Syntactically, the `repr` meta list will be extended to accept a meta item
+name/value pair with the name "align" and the value as a string which can be
+parsed as a `u64`. The restrictions on where this attribute can be placed along
+with the accepted values are:
+
+* Custom alignment can only be specified on `struct` declarations for now.
+  Specifying a different alignment on perhaps `enum` or `type` definitions
+  should be a backwards-compatible extension.
+* Alignment values must be a power of two.
+
+A custom alignment cannot *decrease* the alignment of a structure unless it is
+also declared with `#[repr(packed)]` (to mirror what C does in this regard), but
```

Comment:
Only if you include trailing padding (which we should do something about, at some point).